Bitwise ops for integers

digama0 opened the pull request:
Expansion of the previous commit to include VM implementations of integer bitwise operations.

digama0 committed:
feat(library/vm/vm_nat): implementation of new bitwise ops

gebner commented on library/init/data/int/basic.lean:
```diff
@@ -382,4 +417,10 @@ by rw -int.sub_nat_nat_eq_coe; exact sub_nat_nat_elim m n
   (λi n, by rw [nat.add_sub_cancel_left]; refl)
   (λi n, by rw [add_assoc, nat.sub_eq_zero_of_le (nat.le_add_right _ _)]; refl)
 
+#eval do m ← list.range 5,
```
Debugging code should not be in the library files.

gebner commented on library/init/data/int/bitwise.lean:
```diff
@@ -0,0 +1,63 @@
+  | (m : ℕ) n := nat.test_bit m n
+  | -[1+ m] n := bnot (nat.test_bit m n)
+
+  def nat_bitwise (f : bool → bool → bool) (m n : ℕ) : ℤ :=
```
This function (and `bitwise`) doesn't seem to be used anywhere.  Is there a reason we can't use it for `lor`, `land`, etc.?

digama0 replied:
I'm torn on whether the definitional equality should be the reduction into positive/negative cases as I've done here, or the theorems `bitwise or = lor` etc. Either way we will have both, but definitional reduction can be important here.

gebner replied:
I see, they are used in the theorems at the end of the file.

gebner commented on library/init/data/nat/bitwise.lean:
```diff
@@ -11,15 +11,78 @@ universe u
+
+  def div2 (n : ℕ) : ℕ := (bodd_div2 n).2
+
+  def bodd (n : ℕ) : bool := (bodd_div2 n).1
```
What was wrong with `n % 2 = 1`?

digama0 replied:
Dependence on WF definitions makes it more complicated for the kernel to reduce the definition. I didn't think it would be important, but it kept breaking proofs in weird ways.

gebner replied:
> I didn't think it would be important, but it kept breaking proofs in weird ways.

Do you still remember what it broke?  As far as I can tell, `bodd_div2 n` only reduces if `n` is a numeral.  But in this case I would have expected the WF definition to reduce as well.

digama0 replied:
It's not that it can't reduce; it reduces but only when `unfold_lemmas` is enabled. This causes a problem for  `binary_rec_eq`, because the type of the theorem includes `f ff 0 z = z` (where the left side has type `C (bit ff 0)` and the right side has type `C 0`), so that any users of this theorem must also have `unfold_lemmas` enabled in order to reduce `bit ff 0 := 0`. If they don't, we are in the strange circumstance of having equalities embedded in constants whose type-correctness is not verifiable, and I have no idea what kind of guarantees lean can give in this situation.

gebner replied:
Oooh, I see.  Yes, this is really, really ugly.  And we will probably see this again in the future, since it happens with every WF definition...

I see two potential solutions:
1. Use equational rfl-lemmas for reduction in the type_context.  This is a bit of work, and Leo has mentioned that there are lots of issues here.
2. As a quick hack, we could special-case `acc.rec` (and maybe `eq.rec` and `heq.rec`?) in the type_context by reducing the major premise using `transparency_mode::All`.

BTW, we have exactly the same issue with `attribute [irreducible]`.

gebner commented on library/init/data/nat/gcd.lean:
```diff
@@ -0,0 +1,61 @@
+
+/- gcd -/
+
+def gcd.F : Π (y : ℕ), (Π (y' : ℕ), y' < y → nat → nat) → nat → nat
```
Can we define `gcd` using the equation compiler now?

digama0 replied:
Yes. We should probably also define `div` and `mod` using the equation compiler.

gebner commented:
BTW, this PR doesn't build.

digama0 commented:
Errors in the last commit were fixed, and it has been build-tested this time.

digama0 committed:
feat(init/data/int): prove int bitwise ops

gebner commented on library/init/data/nat/bitwise.lean:
```diff
@@ -11,15 +11,78 @@ universe u
+  @[simp] lemma bodd_one : bodd 1 = tt := rfl
+  @[simp] lemma bodd_two : bodd 2 = ff := rfl
+
+  @[simp] def bodd_succ (n : ℕ) : bodd (succ n) = bnot (bodd n) :=
```
`s/def/lemma/g`

gebner commented:
@digama0 This PR still has build errors.

digama0 commented:
I think this PR is ready for merge (or at least ready for others to try to find issues with it) now.

digama0 committed:
refactor(init/data/nat/bitwise): change definitions to avoid WF

The type-correctness of binary_rec_eq (the statement, not the proof) depends on unfolding the embedded well-founded definition of mod. This definition avoids it by using two simpler functions bodd and div2 that reduce well in the kernel.

digama0 committed:
feat(init/data/int): bitwise ops for integers

digama0 committed:
feat(init/data/int): int lemmas, more bitwise theorems

digama0 committed:
refactor(init/data/nat/gcd): define gcd using eqn compiler

digama0 committed:
fix(tests/lean): fix tests

digama0 commented:
@gebner Is the "request changes" finished?

gebner commented:
> Is the "request changes" finished?

I still think that bodd_div2 is hideous, but I get why we need it.

We should define `gcd` using the equations compiler, but that can be done later.  I don't think we can use the equations compiler for `div` and `mod` for bootstrapping reasons, since they are required by the tactics framework (more precisely `init.data.repr`) and the well-founded recursion support requires tactics.

digama0 commented:
@gebner Actually the latest version of this PR does define `gcd` using the equation compiler. It did involve changing the definition so that it did induction on the left argument instead of the right, but I suppose `gcd` is commutative anyway so all the proofs are simply left-right mirrored.

The other reason for `bodd_div2`, which would remain even if WF definitions worked perfectly, is that it has different definitional equalities. I wanted `bodd (succ n) = bnot (bodd n)` to be a definitional equality instead of whatever you get with the mod definition.

leodemoura commented:
Merged based on @gebner approval.
